### PR TITLE
rpc: add support for unix sockets

### LINKED ISSUE #1224: rpc: add support for unix sockets
Currently, we only support binding to TCP interfaces for both RPC servers. We can easily add support for `unix` sockets as well. This would entail being able to recognize, parse, and then bind to the specified unix socket. 

- [ ] A good starting point would this [this comment that more or less outlines the parsing and binding process](https://github.com/lightningnetwork/lnd/pull/495#issuecomment-355931289).

### PR DESCRIPTION
**Pull request by guggero**
This PR fixes #1224.

Thanks a lot to @Richard87 for [his initial proposal](https://github.com/lightningnetwork/lnd/pull/495#issuecomment-355931289), I was able to use quite a lot of that code.

As described in the issue, Unix sockets are only supported for the gRPC and REST listeners. It doesn't make sense for the p2p listener to listen on an Unix socket.

Currently, TLS cannot be disabled on any of the servers, not even for the loopback interface. So the same applies for Unix sockets, they still require a client certificate to connect.

### CODE REVIEW

File: cmd/lncli/main.go [hunk 1]
```diff
@@ -26,6 +26,8 @@ import (
 const (
 	defaultTLSCertFilename  = "tls.cert"
 	defaultMacaroonFilename = "admin.macaroon"
+	defaultRpcPort          = "10009"
```
Comment (cfromknecht) [inline cmd/lncli/main.go]: noice 👍

File: lnrpc/address.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+			return nil, err
+		}
+		return net.DialTimeout(parsedAddr.Network(),
+			parsedAddr.String(), timeout)
+	}
+}
+
+// IsLoopback returns true if an address describes a loopback interface.
+func IsLoopback(addr string) bool {
+	loopBackAddrs := []string{"localhost", "127.0.0.1", "[::1]"}
```
Comment (cfromknecht) [inline lnrpc/address.go]: We can initialize this at the package level since it's static

File: lnrpc/address.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	return result, nil
+}
+
+// EnforceSafeAuthentication enforces "safe" authentication taking into account
+// the interfaces that the RPC servers are listening on, and if macaroons are
+// activated or not. To protect users from using dangerous config combinations,
+// we'll prevent disabling authentication if the sever is listening on a public
+// interface.
+func EnforceSafeAuthentication(addrs []string, macaroonsActive bool) error {
+
```
Comment (cfromknecht) [inline lnrpc/address.go]: nit: extra new line

File: lnrpc/address_test.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		{"localhost:9735", "tcp", "127.0.0.1:9735", true, false},
+		{"localhost", "tcp", "127.0.0.1:1234", true, false},
+		{"unix:///tmp/lnd.sock", "unix", "/tmp/lnd.sock", false, true},
+		{"unix:/tmp/lnd.sock", "unix", "/tmp/lnd.sock", false, true},
+	}
+)
+
+// TestAddresses ensures that all supported address formats can be parsed and
+// normalized correctly.
+func TestAddresses(t *testing.T) {
```
Comment (cfromknecht) [inline lnrpc/address_test.go]: Excellent test, thorough test vectors 😎

**Comment by halseth** [comment]
This change was surprisingly big. Can you split into smaller commits if possible? :)

Also agree that the methods moved into `lnrpc` probably should not be. I think the change will work even though they are not moved? Makes sense to do the move in a seperate PR (or at least a seperate commit).

File: lnd.go [hunk 5]
```diff
@@ -545,15 +547,17 @@ func lndMain() error {
 		return err
 	}
 	for _, restEndpoint := range cfg.RESTListeners {
-		listener, err := tls.Listen("tcp", restEndpoint, tlsConf)
+		lis, err := lncfg.TlsListenOnNormalizedAddress(restEndpoint,
```
Comment (cfromknecht) [inline lnd.go]: nit: current styling would prefer this line be broken as:
```golang
        lis, err := lncfg.TlsListenOnNormalizedAddress(
                 restEndpoint, tlsConf,
        )
```
also another instance further down

File: config.go [hunk 6]
```diff
@@ -701,11 +704,13 @@ func loadConfig() (*config, error) {
 	// For each of the RPC listeners (REST+gRPC), we'll ensure that users
 	// have specified a safe combo for authentication. If not, we'll bail
 	// out with an error.
-	err := enforceSafeAuthentication(cfg.RPCListeners, !cfg.NoMacaroons)
+	err := lncfg.EnforceSafeAuthentication(cfg.RPCListeners,
```
Comment (cfromknecht) [inline config.go]: similar comment as below wrt. line wrapping on the calls to `EnforceSafeAuthentcation` and `NormalizeAddresses` in this section :)

**Comment by guggero** [comment]
Right, sorry. I'm still getting used to the short lines... :sweat_smile:
I hope it's ok now.

**Comment by halseth** [comment]
Suggestion for how to structure the commits, to make it easier to review:
1. Make all changes to `config.go`, `lnd.go` and `server.go` without renaming/moving the existing methods
2. Move the methods into `lncfg` <-- to keep the diff minimal I would move only what's necessary (it is moved because it is used in `lncli`, right?)
3. Make changes to `lncli`
4. `sample.conf` changes

This is not a blocker though, but I found it a bit hard to review because it's not easy to see at a glance what is new code and what is just code moved 😄

File: lncfg/address.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+func IsUnix(addr string) bool {
+	return strings.HasPrefix(addr, "unix")
+}
+
+// parseAddressString converts an address in string format to a net.Addr that is
+// compatible with lnd. UDP is not supported because lnd needs reliable
+// connections.
+func parseAddressString(strAddress string,
+	defaultPort string) (net.Addr, error) {
+	var parsedNetwork, parsedAddr string
```
Comment (halseth) [inline lncfg/address.go]: The code in these methods is very dense, could you sprinkle in some comments? 😀

File: lncfg/address.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+	case "tcp", "tcp4", "tcp6":
+		return net.ResolveTCPAddr(parsedNetwork,
+			verifyPort(parsedAddr, defaultPort))
+	case "ip", "ip4", "ip6", "udp", "udp4", "udp6", "unixgram":
+		return nil, fmt.Errorf("only TCP or unix socket " +
+			"addresses are supported: %s", parsedAddr)
+	}
+
+	// There was no network specified, just try to parse as host and port.
+	return net.ResolveTCPAddr("tcp", verifyPort(strAddress, defaultPort))
```
Comment (halseth) [inline lncfg/address.go]: can be put in `default`?

File: lncfg/address_test.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+		{"localhost:9735", "tcp", "127.0.0.1:9735", true, false},
+		{"localhost", "tcp", "127.0.0.1:1234", true, false},
+		{"unix:///tmp/lnd.sock", "unix", "/tmp/lnd.sock", false, true},
+		{"unix:/tmp/lnd.sock", "unix", "/tmp/lnd.sock", false, true},
+	}
+)
+
+// TestAddresses ensures that all supported address formats can be parsed and
+// normalized correctly.
+func TestAddresses(t *testing.T) {
```
Comment (halseth) [inline lncfg/address_test.go]: 👍

File: lncfg/address_test.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+		{"tcp4:127.0.0.1:9735", "tcp", "127.0.0.1:9735", true, false},
+		{"127.0.0.1", "tcp", "127.0.0.1:1234", true, false},
+		{"[::1]", "tcp", "[::1]:1234", true, false},
+		{"::1", "tcp", "[::1]:1234", true, false},
+		{"tcp6://::1", "tcp", "[::1]:1234", true, false},
+		{"tcp6:::1", "tcp", "[::1]:1234", true, false},
+		{"localhost:9735", "tcp", "127.0.0.1:9735", true, false},
+		{"localhost", "tcp", "127.0.0.1:1234", true, false},
+		{"unix:///tmp/lnd.sock", "unix", "/tmp/lnd.sock", false, true},
+		{"unix:/tmp/lnd.sock", "unix", "/tmp/lnd.sock", false, true},
```
Comment (halseth) [inline lncfg/address_test.go]: add some invalid strings in here, to make sure they are handled correctly?

File: lncfg/address.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// ListenOnNormalizedAddress creates a listener that listens on the given
+// address. The address must be in the normalized format that is returned
+// by normalizeAddresses, which is network://address[:port].
+// Valid examples:
+//	tcp://localhost:9735
+//	tcp6://[::1]:9735
+//	unix:///tmp/listen.sock
+func ListenOnNormalizedAddress(strAddress string) (net.Listener, error) {
```
Comment (halseth) [inline lncfg/address.go]: Instead of first needing to normalize the address, then splitting it here (meaning we go from `string` -> `net.Addr` -> `string`) it would make sense to let the "normalization" process just return a list of `net.Addr`s, which could be listened on.

File: server.go [hunk 12]
```diff
@@ -147,7 +148,15 @@ func newServer(listenAddrs []string, chanDB *channeldb.DB, cc *chainControl,
 	var err error
 
 	listeners := make([]net.Listener, len(listenAddrs))
-	for i, addr := range listenAddrs {
+	for i, listenAddr := range listenAddrs {
+		// The address is passed in a normalized format with network
+		// and address part. But since the brontide listener only
+		// supports TCP networks anyway, we only need the address part.
+		_, addr, err := lncfg.SplitNormalizedAddress(listenAddr)
```
Comment (halseth) [inline server.go]: if passing in `net.Addr` this wouldn't be needed.

File: lncfg/address.go [hunk 13]
```diff
@@ -8,6 +8,10 @@ import (
 	"time"
 )
 
+var (
+	loopBackAddrs = []string{"localhost", "127.0.0.1", "[::1]"}
```
Comment (halseth) [inline lncfg/address.go]: belongs in a different commit?

**Comment by vegardengen** [comment]
In general, I think sockets belong in /var/run somewhere and not in /tmp. Would it be a good idea to change the examples?

This, imho, should be gone through for the whole of LND, I would prefer default config file location to be moved to /etc, the data dirs etc to be moved to /var/lib/lnd - and other Unixfying changes.

**Comment by guggero** [comment]
It took me a while but I think I was able to structure the commits as you suggested, @halseth.

@vegardengen: I updated the example. I think with the `--configfile` and `--datadir` options, lnd allows customization so the linux distribution package maintainers can put the files wherever their distro defaults to. In my opinion this isn't necessarily the tool's job since any chosen default will be 'wrong' in some distro's eye...

**Comment by cfromknecht** [comment]
@guggero adding this to the 0.5 milestone, it's ready to go IMO. Will need a rebase and then can start the merge process once we cut 0.4.2 🎉

**Comment by cfromknecht** [comment]
@guggero MERGING SEASON IS HERE. Should be good for a final rebase/review

**Comment by guggero** [comment]
Yeah :sunglasses: 
I've rebased all my open PRs.

**Comment by Roasbeef** [comment]
Needs a rebase, thanks!

**Comment by guggero** [comment]
Rebased :)

**Comment by cfromknecht** [comment]
tests failing, looks like some minor compilation erros

**Comment by guggero** [comment]
Sorry, I made a mistake with the last rebase. The Tor support had more influence on this PR than I thought...

**Comment by Roasbeef** [comment]
Latest travis run: 
```
# github.com/lightningnetwork/lnd
./config.go:845:25: undefined: normalizeAddresses
./server.go:387:22: undefined: normalizeAddresses
```

**Comment by guggero** [comment]
Fixed and rebased

**Comment by Roasbeef** [comment]
Will give this a spin locally, then merge if all seems well!
